File: scripts/update-stats.js
// npm run update-stats

const fs = require('fs');
const path = require('path');

const CACHE_FILE = path.join(__dirname, '..', 'github-stats-cache.json');
const DATA_FILE = path.join(__dirname, '..', 'js', 'data.js');

async function fetchRepoStats(owner, repo, token) {
    const url = `https://api.github.com/repos/${owner}/${repo}`;
    const headers = {
        'User-Agent': 'WowkDigital-WD-HUB-Updater'
    };
    if (token) {
        headers['Authorization'] = `token ${token}`;
    }

    try {
        const response = await fetch(url, { headers });
        if (!response.ok) {
            console.error(`[ERROR] Failed to fetch stats for ${owner}/${repo}: ${response.status} ${response.statusText}`);
            if (response.status === 403 || response.status === 429) {
                return { rateLimited: true, status: response.status };
            }
            return null;
        }
        const data = await response.json();
        return {
            created_at: data.created_at,
            pushed_at: data.pushed_at || data.updated_at,
            updated_at: data.updated_at
        };
    } catch (e) {
        console.error(`[ERROR] Error fetching stats for ${owner}/${repo}:`, e.message);
        return null;
    }
}

async function run() {
    console.log('Loading projects from data.js...');
    if (!fs.existsSync(DATA_FILE)) {
        console.error(`[ERROR] Data file not found at: ${DATA_FILE}`);
        process.exit(1);
    }

    const { projects } = require(DATA_FILE);
    if (!projects || !Array.isArray(projects)) {
        console.error('[ERROR] Invalid projects structure in data.js');
        process.exit(1);
    }

    // Load existing cache if exists to keep on failure
    let existingCache = { data: {} };
    if (fs.existsSync(CACHE_FILE)) {
        try {
            existingCache = JSON.parse(fs.readFileSync(CACHE_FILE, 'utf8'));
        } catch (e) {
            // Ignore parse errors
        }
    }

    const token = process.env.GITHUB_TOKEN;
    if (!token) {
        console.log('[WARN] No GITHUB_TOKEN environment variable found. Rate limits may apply.');
    } else {
        console.log('[INFO] Using GitHub Token for authenticated API calls.');
    }

    const newData = {};
    let rateLimited = false;
    console.log(`Starting update for ${projects.length} projects...`);

    for (let i = 0; i < projects.length; i++) {
        const project = projects[i];
        if (!project.github) {
            console.log(`[INFO] [${i + 1}/${projects.length}] Skipping "${project.title}" (no GitHub URL)`);
            continue;
        }

        if (rateLimited) {
            if (existingCache.data[project.github]) {
                newData[project.github] = existingCache.data[project.github];
            }
            continue;
        }

        const match = project.github.match(/github\.com\/([^\/]+)\/([^\/]+)/);
        if (match) {
            const owner = match[1];
            const repo = match[2].replace(/.git$/, '');
            console.log(`[FETCH] [${i + 1}/${projects.length}] Fetching stats for ${owner}/${repo}...`);

            const stats = await fetchRepoStats(owner, repo, token);
            if (stats && stats.rateLimited) {
                console.log(`   [RATE LIMIT] GitHub API Rate limit hit. Suspending further API requests.`);
                rateLimited = true;
                if (existingCache.data[project.github]) {
                    newData[project.github] = existingCache.data[project.github];
                }
                continue;
            }

            if (stats) {
                newData[project.github] = stats;
                console.log(`   [SUCCESS] Created: ${stats.created_at.split('T')[0]} | Updated: ${stats.pushed_at.split('T')[0]}`);
            } else if (existingCache.data[project.github]) {
                newData[project.github] = existingCache.data[project.github];
                console.log(`   [WARN] Kept existing cached stats.`);
            } else {
                console.log(`   [FAIL] Failed to get stats.`);
            }

            // Sleep 500ms to avoid hitting rate limit
            await new Promise(resolve => setTimeout(resolve, 500));
        } else {
            console.log(`[WARN] [${i + 1}/${projects.length}] Invalid GitHub URL format: ${project.github}`);
        }
    }

    const finalCache = {
        timestamp: Date.now(),
        data: newData
    };

    fs.writeFileSync(CACHE_FILE, JSON.stringify(finalCache, null, 2), 'utf8');
    console.log(`\nStats update completed! Saved to ${CACHE_FILE}`);
}

run();

File: scripts/update_stats.py
import os
import json
import time
import urllib.request
import urllib.error
import re
import sys

# Ensure stdout supports UTF-8, especially on Windows terminal
if hasattr(sys.stdout, 'reconfigure'):
    try:
        sys.stdout.reconfigure(encoding='utf-8')
    except Exception:
        pass

CACHE_FILE = "github-stats-cache.json"
DATA_FILE = os.path.join("js", "data.js")

def get_repo_details(owner, repo, token):
    url = f"https://api.github.com/repos/{owner}/{repo}"
    headers = {'User-Agent': 'WowkDigital-WD-HUB-Updater'}
    if token:
        headers['Authorization'] = f"token {token}"
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read().decode())
            return {
                "created_at": data.get("created_at"),
                "pushed_at": data.get("pushed_at") or data.get("updated_at"),
                "updated_at": data.get("updated_at")
            }
    except urllib.error.HTTPError as e:
        print(f"[ERROR] Failed to fetch stats for {owner}/{repo}: {e.code} {e.reason}")
        if e.code in [403, 429]:
            return {"rateLimited": True, "status": e.code}
        return None
    except Exception as e:
        print(f"[ERROR] Error fetching stats for {owner}/{repo}: {e}")
        return None

def parse_projects_from_js():
    if not os.path.exists(DATA_FILE):
        print(f"[ERROR] Data file not found at: {DATA_FILE}")
        sys.exit(1)
        
    with open(DATA_FILE, "r", encoding="utf-8") as f:
        content = f.read()
        
    # Extract projects array contents
    # Looking for: const projects = [ ... ];
    match = re.search(r"const\s+projects\s*=\s*\[([\s\S]*?)\];", content)
    if not match:
        print("[ERROR] Could not find projects list in js/data.js")
        sys.exit(1)
        
    projects_content = match.group(1)
    
    # We will use simple regexes to find title and github fields within the JS objects
    # Splitting into individual object strings
    objects = re.findall(r"\{([\s\S]*?)\}", projects_content)
    
    projects = []
    for obj in objects:
        title_match = re.search(r'title:\s*["\'](.*?)["\']', obj)
        github_match = re.search(r'github:\s*["\'](.*?)["\']', obj)
        
        if title_match:
            title = title_match.group(1)
            github = github_match.group(1) if github_match else None
            projects.append({"title": title, "github": github})
            
    return projects

def main():
    print("Loading projects from data.js...")
    projects = parse_projects_from_js()
    
    # Load existing cache
    existing_cache = {"data": {}}
    if os.path.exists(CACHE_FILE):
        try:
            with open(CACHE_FILE, "r", encoding="utf-8") as f:
                existing_cache = json.load(f)
        except Exception:
            pass
            
    if "data" not in existing_cache:
        existing_cache["data"] = {}
        
    token = os.environ.get('GITHUB_TOKEN') or os.environ.get('GH_TOKEN')
    if not token:
        print("[WARN] No GITHUB_TOKEN environment variable found. Rate limits may apply.")
    else:
        print("[INFO] Using GitHub Token for authenticated API calls.")
        
    new_data = {}
    rate_limited = False
    print(f"Starting update for {len(projects)} projects...")
    
    for i, project in enumerate(projects):
        github_url = project.get("github")
        title = project.get("title")
        
        if not github_url:
            print(f"[INFO] [{i + 1}/{len(projects)}] Skipping \"{title}\" (no GitHub URL)")
            continue
            
        github_url = github_url.strip().rstrip('/')
        if rate_limited:
            if github_url in existing_cache["data"]:
                new_data[github_url] = existing_cache["data"][github_url]
            continue
            
        parts = github_url.split('/')
        if len(parts) >= 5 and "github.com" in parts[2]:
            owner = parts[3]
            repo = parts[4]
            print(f"[FETCH] [{i + 1}/{len(projects)}] Fetching stats for {owner}/{repo}...")
            
            stats = get_repo_details(owner, repo, token)
            if stats and stats.get("rateLimited"):
                print("   [RATE LIMIT] GitHub API Rate limit hit. Suspending further API requests.")
                rate_limited = True
                if github_url in existing_cache["data"]:
                    new_data[github_url] = existing_cache["data"][github_url]
                continue
                
            if stats:
                new_data[github_url] = stats
                print(f"   [SUCCESS] Created: {stats['created_at'].split('T')[0]} | Updated: {stats['pushed_at'].split('T')[0]}")
            elif github_url in existing_cache["data"]:
                new_data[github_url] = existing_cache["data"][github_url]
                print("   [WARN] Kept existing cached stats.")
            else:
                print("   [FAIL] Failed to get stats.")
                
            time.sleep(0.5)
        else:
            print(f"[WARN] [{i + 1}/{len(projects)}] Invalid GitHub URL format: {github_url}")
            
    final_cache = {
        "timestamp": int(time.time() * 1000),
        "data": new_data
    }
    
    try:
        with open(CACHE_FILE, "w", encoding="utf-8") as f:
            json.dump(final_cache, f, indent=2)
        print(f"\nStats update completed! Saved to {CACHE_FILE}")
    except Exception as e:
        print(f"[ERROR] Failed to save cache: {e}")

if __name__ == "__main__":
    main()
